### ISSUE TITLE
crash of lnst-slave in _rpc_call_to_netns

### ISSUE THREAD

**Issue by jtluka** [jpirko/lnst#99]
Following traceback while running a recipe with namespaces involved.
```
    Traceback (most recent call last):
      File "/usr/bin/lnst-ctl", line 122, in get_recipe_result
        res = exec_action(action, nettestctl)
      File "/usr/bin/lnst-ctl", line 79, in exec_action
        return nettestctl.run_recipe()
      File "/usr/lib/python2.6/site-packages/lnst/Controller/NetTestController.py", line 631, in run_recipe
        res = self._run_recipe()
      File "/usr/lib/python2.6/site-packages/lnst/Controller/NetTestController.py", line 649, in _run_recipe
        res = self._run_task(task)
      File "/usr/lib/python2.6/site-packages/lnst/Controller/NetTestController.py", line 694, in _run_task
        return self._run_python_task(task)
      File "/usr/lib/python2.6/site-packages/lnst/Controller/NetTestController.py", line 684, in _run_python_task
        module = imp.load_source(name, task_path.resolve())
      File "/root/ipsec-lnst-tests2/ipsec-comp-netns.py", line 130, in <module>
        dump.intr()
      File "/usr/lib/python2.6/site-packages/lnst/Controller/Task.py", line 388, in intr
        self._res = self._ctl._run_command(cmd)
      File "/usr/lib/python2.6/site-packages/lnst/Controller/Task.py", line 36, in _run_command
        res = self._ctl._run_command(command)
      File "/usr/lib/python2.6/site-packages/lnst/Controller/NetTestController.py", line 726, in _run_command
        cmd_res = machine.run_command(command)
      File "/usr/lib/python2.6/site-packages/lnst/Controller/Machine.py", line 290, in run_command
        cmd_res = self._rpc_call_to_netns(netns, "run_command", command)
      File "/usr/lib/python2.6/site-packages/lnst/Controller/Machine.py", line 163, in _rpc_call_to_netns
        result = self._msg_dispatcher.wait_for_result(self._id)
      File "/usr/lib/python2.6/site-packages/lnst/Controller/NetTestController.py", line 874, in wait_for_result
        raise NetTestError(msg)
    NetTestError: Slaves ['machine1'] disconnected from the controller.
```

**Comment by jtluka** [comment]
Attaching lnst recipe and python task.

file ipsec-comp-netns.xml
```xml
<lnstrecipe>
    <network>
        <host id="machine1">
            <interfaces>
                <eth id="out_if" label="localnet">
                    <addresses>
                        <address>192.168.10.1/24</address>
                    </addresses>
                </eth>
                <veth_pair>
                    <veth id="testifc1" netns="xyz" >
                        <addresses>
                            <address>192.168.10.3/24</address>
                        </addresses>
                    </veth>
                    <veth id="testifc2" />
                </veth_pair>
                <bridge id="br">
                    <slaves>
                        <slave id="out_if" />
                        <slave id="testifc2" />
                    </slaves>
                </bridge>

            </interfaces>
        </host>
        <host id="machine2">
            <interfaces>
                <eth id="testiface" label="localnet">
                    <addresses>
                        <address>192.168.10.2/24</address>
                    </addresses>
                </eth>
            </interfaces>
        </host>

    </network>

    <task>
        <run host="machine1" module="IcmpPing" netns="xyz">
            <options>
                <option name="addr" value="{ip(machine2, testiface)}" />
                <option name="count" value = "3" />
                <option name="iface" value="{devname(machine1, testifc1)}" />
            </options>
        </run>
    </task>
    <task python="ipsec-comp-netns.py" />
</lnstrecipe>
```

file ipsec-comp-netns.py

```python
from lnst.Controller.Task import ctl

m1 = ctl.get_host("machine1")
m2 = ctl.get_host("machine2")

m1.sync_resources(modules=["IcmpPing"])

m2.sync_resources(modules=["PacketAssert"])

m1_addr = m1.get_ip("testifc1")

m2_addr = m2.get_ip("testiface")

res = m1.run("rpm -qa iproute", save_output=True)
if (res.get_result()["res_data"]["stdout"].find("iproute-2") != -1):
    m1_key="0x"
else:
    m1_key=""

res = m2.run("rpm -qa iproute", save_output=True)
if (res.get_result()["res_data"]["stdout"].find("iproute-2") != -1):
    m2_key="0x"
else:
    m2_key=""


# ------
# TESTS

# lzs and lzjh are not working on rhel6
for comp_alg in [ 'deflate' ]:
    # configure policy and state
    m1.run("ip xfrm policy flush", netns="xyz")
    m1.run("ip xfrm state flush", netns="xyz")

    res = m1.run("ip xfrm policy add dir out src %s dst %s "\
                 "tmpl src %s dst %s proto comp spi 2 mode tunnel "\
                 "tmpl src %s dst %s proto esp spi 1 mode tunnel"\
                  % (m1_addr, m2_addr, m1_addr, m2_addr, m1_addr, m2_addr), netns="xyz")
    res = m1.run("ip xfrm policy add dir in src %s dst %s "\
                 "tmpl src %s dst %s proto comp spi 4 mode tunnel "\
                 "tmpl src %s dst %s proto esp spi 3 mode tunnel"\
                  % (m2_addr, m1_addr, m2_addr, m1_addr, m2_addr, m1_addr), netns="xyz")

    if not res.passed():
        m1.run("exit 1", desc="ip xfrm failed on alg: %s" % comp_alg, netns="xyz")
        continue

    res = m1.run("ip xfrm state add src %s dst %s "\
                 "proto comp spi 2 mode tunnel comp %s %s" % (m1_addr, m2_addr, comp_alg, m1_key), netns="xyz")
    res = m1.run("ip xfrm state add src %s dst %s "\
                 "proto comp spi 4 mode tunnel comp %s %s" % (m2_addr, m1_addr, comp_alg, m1_key), netns="xyz")
    res = m1.run("ip xfrm state add src %s dst %s "\
                 "proto esp spi 1 mode tunnel "\
                 "enc des3_ede 0x112233445566778811223344556677881122334455667788"\
                  % (m1_addr, m2_addr), netns="xyz")
    res = m1.run("ip xfrm state add src %s dst %s "\
                 "proto esp spi 3 mode tunnel "\
                 "enc des3_ede 0x112233445566778811223344556677881122334455667788"\
                  % (m2_addr, m1_addr), netns="xyz")

    if not res.passed():
        m1.run("exit 1", desc="ip xfrm failed on alg: %s" % comp_alg, netns="xyz")
        continue

    m2.run("ip xfrm policy flush")
    m2.run("ip xfrm state flush")

    res = m2.run("ip xfrm policy add dir out src %s dst %s "\
                 "tmpl src %s dst %s proto comp spi 4 mode tunnel "\
                 "tmpl src %s dst %s proto esp spi 3 mode tunnel"\
                  % (m2_addr, m1_addr, m2_addr, m1_addr, m2_addr, m1_addr))
    res = m2.run("ip xfrm policy add dir in src %s dst %s "\
                 "tmpl src %s dst %s proto comp spi 2 mode tunnel "\
                 "tmpl src %s dst %s proto esp spi 1 mode tunnel"\
                  % (m1_addr, m2_addr, m1_addr, m2_addr, m1_addr, m2_addr))
    if not res.passed():
        m2.run("exit 1", desc="ip xfrm failed on alg: %s" % comp_alg)
        continue

    res = m2.run("ip xfrm state add src %s dst %s "\
                 "proto comp spi 2 mode tunnel comp %s %s" %(m1_addr, m2_addr, comp_alg, m2_key))
    res = m2.run("ip xfrm state add src %s dst %s "\
                 "proto comp spi 4 mode tunnel comp %s %s" %(m2_addr, m1_addr, comp_alg, m2_key))
    res = m2.run("ip xfrm state add src %s dst %s "\
                 "proto esp spi 1 mode tunnel "\
                 "enc des3_ede 0x112233445566778811223344556677881122334455667788"\
                  % (m1_addr, m2_addr))
    res = m2.run("ip xfrm state add src %s dst %s "\
                 "proto esp spi 3 mode tunnel "\
                 "enc des3_ede 0x112233445566778811223344556677881122334455667788"\
                  % (m2_addr, m1_addr))

    if not res.passed():
        m2.run("exit 1", desc="ip xfrm failed on alg: %s" % comp_alg)
        continue

    m1.run("ip xfrm policy", netns="xyz")
    m1.run("ip xfrm state", netns="xyz")
    m2.run("ip xfrm policy")
    m2.run("ip xfrm state")

    dump = m1.run("tcpdump -i %s -vv" % m1.get_devname("testifc1"), netns="xyz", bg=True)

    # ping + PacketAssert
    assert_mod = ctl.get_module("PacketAssert",
                             options={
                                 "interface": m2.get_devname("testiface"),
                                 "filter": "ip proto ipcomp",
                                 "min": 5
                             })

    assert_proc = m2.run(assert_mod, bg=True)

    ping_mod=ctl.get_module("IcmpPing",
                            options={
                                "addr": m2_addr,
                                "count": 8,
                                "interval": 0.2,
                                "iface" : m1.get_devname("testifc1"),
                                "size": 1000})

    ctl.wait(2)

    m1.run(ping_mod, netns="xyz")

    ctl.wait(2)

    assert_proc.intr()
    dump.intr()

m1.run("ip xfrm pol flush", netns="xyz")
m1.run("ip xfrm state flush", netns="xyz")
m2.run("ip xfrm pol flush")
m2.run("ip xfrm state flush")
```

**Comment by jtluka** [comment]
Seems that this is a regression introduced by commit 110ab74f9c4f9a60fd42d029a5ea67e04af3f3b9 
NetTestSlave: signal handling for lnst-slave in netns

**Comment by jtluka** [comment]
lnst-slave counter-part traceback:
```
Traceback (most recent call last):
  File "/bin/lnst-slave", line 98, in <module>
    main()
  File "/bin/lnst-slave", line 95, in main
    nettestslave.run()
  File "/usr/lib/python2.7/site-packages/lnst/Slave/NetTestSlave.py", line 773, in run
    self._methods.machine_cleanup()
  File "/usr/lib/python2.7/site-packages/lnst/Slave/NetTestSlave.py", line 383, in machine_cleanup
    self.deconfigure_if_pair(if_id, peer_id)
  File "/usr/lib/python2.7/site-packages/lnst/Slave/NetTestSlave.py", line 230, in deconfigure_if_pair
    raise Exception("Deconfiguration failed.")
Exception: Deconfiguration failed.
Traceback (most recent call last):
  File "/bin/lnst-slave", line 98, in <module>
    main()
  File "/bin/lnst-slave", line 95, in main
    nettestslave.run()
  File "/usr/lib/python2.7/site-packages/lnst/Slave/NetTestSlave.py", line 768, in run
    msgs = self._server_handler.get_messages()
  File "/usr/lib/python2.7/site-packages/lnst/Slave/NetTestSlave.py", line 659, in get_messages
    messages = self.check_connections()
  File "/usr/lib/python2.7/site-packages/lnst/Common/ConnectionHandler.py", line 74, in check_connections
    return self._check_connections(self._connections)
  File "/usr/lib/python2.7/site-packages/lnst/Common/ConnectionHandler.py", line 92, in _check_connections
    data = recv_data(f)
  File "/usr/lib/python2.7/site-packages/lnst/Common/ConnectionHandler.py", line 62, in recv_data
    data = s.recv()
IOError: [Errno 104] Connection reset by peer
```

**Comment by jtluka** [comment]
Here's the output from lnst-slave when signal handling happens on intr()

2015-02-05 19:02:53       (localhost)      xyz   DEBUG: Interrupting background command with id "1", pid "16783"
2015-02-05 19:02:53       (localhost)        -    INFO: Caught signal 2 -> dying
2015-02-05 19:02:53       (localhost)        -    INFO: Performing machine cleanup.
2015-02-05 19:02:53       (localhost)        -   DEBUG: Cleaning up leftover processes.
2015-02-05 19:02:53       (localhost)        -   DEBUG: Executing: "ip link set t_br0 down"
2015-02-05 19:02:53       (localhost)        -   DEBUG: Executing: "brctl delif t_br0 ens30"
2015-02-05 19:02:53       (localhost)        -   DEBUG: Executing: "brctl delif t_br0 veth1"
2015-02-05 19:02:53       (localhost)        -   DEBUG: Executing: "brctl delbr t_br0"
2015-02-05 19:02:53       (localhost)        -   DEBUG: Executing: "ip addr del 192.168.10.1/24 dev ens30"
2015-02-05 19:02:53       (localhost)        -   DEBUG: Executing: "ip link set ens30 down"
2015-02-05 19:02:53       (localhost)      xyz    INFO: Caught signal 2 -> resending to parent
2015-02-05 19:02:53       (localhost)      xyz    INFO: Caught signal 2 -> resending to parent

**Comment by jtluka** [comment]
The problem is that intr() actually emits SIGINT on the whole process group,
```os.killpg(os.getpgid(self._pid), signal.SIGINT)```
and the signal handler from netns process resends to the main slave server process, that ends the whole lnst-slave program.

**Comment by jtluka** [comment]
Here's the pstree output from the command,
```
           |-sshd(784)---sshd(17929)---bash(17932)---lnst-slave(18634)---lnst-slave(18649)-+-lnst-slave(18738)---tcpdump(18739)
           |                                                                               `-pstree(18742)
```

And some related debugging output,
```
18738
2015-02-09 15:02:30       (localhost)      xyz   DEBUG: Interrupting background command with id "1", pid "18738"
2015-02-09 15:02:30       (localhost)      xyz   DEBUG: the pgpid is == 18738
```

So, it seems that we're actually killing the whole netns-forked lnst-slave process, not the tcpdump command (the one that we call intr() on)

**Comment by jtluka** [comment]
More likely that I was wrong with my previous comment, the last lnst-slave process in pstree output is actually the forked part that spawns NetTestCommands

**Comment by jtluka** [comment]
So, together with Ondrej we've finally found out what's causing this.

If a netns device is in the recipe we do fork() and set signal handlers in child process. If we start a background process in the netns, we do another fork() and signal handlers are INHERITED again. So there's actually a race when bg_process.intr() comes in. The solution is to remove signal handlers in the NetTestCommand._run().

**Comment by olichtne** [comment]
Fixed in commit 7b7a88a54b79c812a711bc05de6e2a5956aee784
